woocommerce/woocommerce
# feat: add `aria-required` attributes to WC form fields

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Adding the `aria-required="true"` attribute to required form inputs.
I chose to add the `aria-required` attribute rather than the `required` attribute to avoid backwards-compatibility issues.
When the `required` attribute is added, the form submission is intercepted and prevented by the browser.
By adding `aria-required`, instead`, we add information to Screen Reader users without affecting the form's behavior.

I ensured that the `aria-required` attribute was added to all input types, excluding `hidden`.
The `aria-required` attribute is added to all inputs of type `radio`, [per HTML specs](https://html.spec.whatwg.org/multipage/input.html#the-required-attribute):
> To avoid confusion as to whether a [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) is required or not, authors are encouraged to specify the attribute on all the radio buttons in a group. 





Closes https://github.com/woocommerce/woocommerce/issues/44663



### How to test the changes in this Pull Request:



- (optional) Add the following snippet to your installation to test the radio inputs(you can use [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/), for example)
```php
add_action( 'woocommerce_after_checkout_billing_form', 'checkout_test_radio_options' );
function checkout_test_radio_options() {
	echo "<div>";
	echo sprintf( '<p>%s</p>', __( "Test radio input" ) );
    woocommerce_form_field( 'test_radio',  array(
		'required'  => true,
        'type'      => 'radio',
        'class'     => array( 'form-row-wide' ),
        'options'   => array(
            '1' => 'Option 1',
            '2' => 'Option 2',
        ),
    ), '1' );
	echo "</div>";
}
```
- Add a product to the cart
- Go to the shortcode checkout page
- The checkout page should display a wide range of inputs. All required inputs (excluding the "hidden" ones) should have an `aria-required="true"` attribute



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
feat: add `aria-required` attributes to WC form fields
</details>


<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
